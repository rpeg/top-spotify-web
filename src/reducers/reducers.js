import { combineReducers } from 'redux';
import {
  SET_USER,
  SET_TIME_RANGE,
  TimeRanges,
  REQUEST_ARTISTS,
  RECEIVE_ARTISTS,
  REQUEST_TRACKS,
  RECEIVE_TRACKS,
  REQUEST_FEATURES,
  RECEIVE_FEATURES,
  SET_ARTIST_COUNT,
  SET_TRACK_COUNT,
  SET_GENRE_COUNT,
} from '../constants/constants';

const { LONG } = TimeRanges;

function user(state = {}, action) {
  switch (action.type) {
    case SET_USER:
      return { ...state, user: action.user };
    default:
      return state;
  }
}

function timeRange(state = LONG, action) {
  switch (action.type) {
    case SET_TIME_RANGE:
      return { ...state, timeRange: action.timeRange };
    default:
      return state;
  }
}

function artistCount(state = 10, action) {
  switch (action.type) {
    case SET_ARTIST_COUNT:
      return { ...state, artistCount: action.artistCount };
    default:
      return state;
  }
}

function trackCount(state = 15, action) {
  switch (action.type) {
    case SET_TRACK_COUNT:
      return { ...state, trackCount: action.trackCount };
    default:
      return state;
  }
}

function genreCount(state = 40, action) {
  switch (action.type) {
    case SET_GENRE_COUNT:
      return { ...state, genreCount: action.genreCount };
    default:
      return state;
  }
}

function artists(state = {}, action) {
  switch (action.type) {
    case REQUEST_ARTISTS:
      return { ...state, isFetching: true };
    case RECEIVE_ARTISTS:
      return { ...state, isFetching: false, items: action.items };
    default:
      return state;
  }
}

function artistsByTimeRange(state = {}, action) {
  switch (action.type) {
    case RECEIVE_ARTISTS:
    case REQUEST_ARTISTS:
      return { ...state, [action.timeRange]: artists(state[action.timeRange], action) };
    default:
      return state;
  }
}

function tracks(state = {}, action) {
  switch (action.type) {
    case REQUEST_TRACKS:
      return { ...state, isFetching: true };
    case RECEIVE_TRACKS:
      return { ...state, isFetching: false, items: action.tracks };
    default:
      return state;
  }
}

function tracksByTimeRange(state = {}, action) {
  switch (action.type) {
    case RECEIVE_TRACKS:
    case REQUEST_TRACKS:
      return { ...state, [action.timeRange]: tracks(state[action.timeRange], action) };
    default:
      return state;
  }
}

function features(state = {}, action) {
  switch (action.type) {
    case REQUEST_FEATURES:
      return { ...state, isFetching: true };
    case RECEIVE_FEATURES:
      return { ...state, isFetching: false, items: action.features };
    default:
      return state;
  }
}

const rootReducer = combineReducers({
  user,
  timeRange,
  artistCount,
  trackCount,
  genreCount,
  artists,
  artistsByTimeRange,
  tracks,
  tracksByTimeRange,
  features,
});

export default rootReducer;
