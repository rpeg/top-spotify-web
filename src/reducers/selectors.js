export const selectHasClickedCreate = (state) => state.hasClickedCreate;
export const selectUser = (state) => state.user;
export const selectTimeRangeName = (state) => state.timeRangeName;
export const selectGenreCount = (state) => state.genreCount;
export const selectArtistCount = (state) => state.artistCount;
export const selectTrackCount = (state) => state.trackCount;
export const selectStatsOptions = (state) => state.statsOptions;
export const selectDisplayProfile = (state) => state.displayProfile;
export const selectOptimizeTracks = (state) => state.optimizeTracks;
export const selectArtists = (state) => state.artists;
export const selectArtistsByCurrentTimeRange = (state) => state.artistsByTimeRangeName[state.timeRangeName];
export const selectTracks = (state) => state.tracks;
export const selectTracksByCurrentTimeRange = (state) => state.tracksByTimeRangeName[state.timeRangeName];
export const selectFeatures = (state) => state.features;
export const selectFeaturesByCurrentTimeRange = (state) => state.featuresByTimeRangeName[state.timeRangeName];
