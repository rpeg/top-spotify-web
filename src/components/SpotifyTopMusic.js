import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Spinner, Container } from 'react-bootstrap';

import ArtistGrid from './ArtistGrid';
import TrackGrid from './TrackGrid';
import WordCloud from './WordCloud';
import Statistics from './Statistics';
import ComponentHeader from './ComponentHeader';
import {
  fetchArtistsIfNeeded,
  fetchTracksIfNeeded,
} from '../actions/actions';
import { getTimeRangeByName } from '../lib/timeRange';
import * as selectors from '../reducers/selectors';

const SpinnerBlock = () => (
  <div style={{ marginTop: '2em' }}>
    <Spinner animation="border" />
  </div>
);

const SpotifyTopMusic = () => {
  const user = useSelector(selectors.selectUser);
  const timeRangeName = useSelector(selectors.selectTimeRangeName);
  const artistCount = useSelector(selectors.selectArtistCount);
  const trackCount = useSelector(selectors.selectTrackCount);
  const genreCount = useSelector(selectors.selectGenreCount);
  const statsOptions = useSelector(selectors.selectStatsOptions);
  const artists = useSelector(selectors.selectArtistsByCurrentTimeRange);
  const tracks = useSelector(selectors.selectTracksByCurrentTimeRange);
  const features = useSelector(selectors.selectFeaturesByCurrentTimeRange);

  const dispatch = useDispatch();

  useEffect(() => {
    if (user && user.id) {
      dispatch(fetchArtistsIfNeeded(timeRangeName));
      dispatch(fetchTracksIfNeeded(timeRangeName));
    }
  }, [dispatch, timeRangeName, user]);

  const haveArtists = () => artists && artists.items && artists.items.length;
  const haveTracks = () => tracks && tracks.items && tracks.items.length;
  const haveFeatures = () => features && features.items && features.items.length;

  return (
    user && user.id ? (
      <div style={{ padding: '2em 0 0 0' }}>
        <div>
          <p className="text-xl" style={{ marginBottom: '0uu' }}>{getTimeRangeByName(timeRangeName).title}</p>
        </div>
        <Container>
          {artistCount > 0 && (haveArtists()
            ? (
              <div>
                <ComponentHeader title="Artists" />
                <div style={{ marginTop: '1em' }}>
                  <ArtistGrid artists={artists.items.slice(0, artistCount)} />
                </div>
              </div>
            )
            : <SpinnerBlock />)}
          {genreCount > 0 && (haveArtists()
            ? (
              <div>
                <ComponentHeader title="Genres" />
                <WordCloud
                  genres={[...artists.items.map((a) => a.genres).flat()]}
                  count={genreCount}
                />
              </div>
            )
            : <SpinnerBlock />)}
          {trackCount > 0 && (haveTracks()
            ? (
              <div>
                <ComponentHeader title="Tracks" />
                <TrackGrid tracks={tracks.items} count={trackCount} />
              </div>
            )
            : <SpinnerBlock />)}
          {statsOptions.length > 0 && (haveTracks() && haveFeatures()
            ? (
              <div>
                <ComponentHeader title="Statistics" />
                <Statistics features={features.items} tracks={tracks.items} />
              </div>
            )
            : <SpinnerBlock />)}
        </Container>
      </div>
    ) : null
  );
};

export default SpotifyTopMusic;
