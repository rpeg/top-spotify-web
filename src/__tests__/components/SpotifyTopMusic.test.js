import React from 'react';
import { shallow } from 'enzyme';

import SpotifyTopMusic from '../../components/SpotifyTopMusic';
import * as selectors from '../../reducers/selectors';
import {
  TimeRanges, DEFAULT_GENRE_COUNT, DEFAULT_ARTIST_COUNT, DEFAULT_TRACK_COUNT, DEFAULT_STATS_OPTIONS,
} from '../../constants/constants';

jest.mock('react-redux', () => ({
  useSelector: jest.fn((fn) => fn()),
  useDispatch: () => jest.fn(),
}));

const setup = (
  user = {},
  timeRangeName = TimeRanges.LONG.name,
  genreCount = DEFAULT_GENRE_COUNT,
  artistCount = DEFAULT_ARTIST_COUNT,
  trackCount = DEFAULT_TRACK_COUNT,
  statsOptions = DEFAULT_STATS_OPTIONS,
  artistsByTimeRangeName = {},
  tracksByTimeRangeName = {},
  featuresByTimeRangeName = {},
  artistCountriesByTimeRangeName = {},
  artistReleasesByTimeRangeName = {},
) => {
  jest.spyOn(selectors, 'selectUser').mockReturnValue(user);
  jest.spyOn(selectors, 'selectTimeRangeName').mockReturnValue(timeRangeName);
  jest.spyOn(selectors, 'selectGenreCount').mockReturnValue(genreCount);
  jest.spyOn(selectors, 'selectArtistCount').mockReturnValue(artistCount);
  jest.spyOn(selectors, 'selectTrackCount').mockReturnValue(trackCount);
  jest.spyOn(selectors, 'selectStatsOptions').mockReturnValue(statsOptions);
  jest.spyOn(selectors, 'selectArtistsByCurrentTimeRange').mockReturnValue(artistsByTimeRangeName);
  jest.spyOn(selectors, 'selectTracksByCurrentTimeRange').mockReturnValue(tracksByTimeRangeName);
  jest.spyOn(selectors, 'selectFeaturesByCurrentTimeRange').mockReturnValue(featuresByTimeRangeName);
  jest.spyOn(selectors, 'selectArtistCountriesByCurrentTimeRange').mockReturnValue(artistCountriesByTimeRangeName);
  jest.spyOn(selectors, 'selectArtistReleasesByCurrentTimeRange').mockReturnValue(artistReleasesByTimeRangeName);
};

describe('SpotifyTopMusic', () => {
  beforeEach(() => {
    setup(
      { id: 'abc' },
      TimeRanges.LONG.name,
      10,
      10,
      10,
      ['bpm'],
      {
        LONG: [],
      },
      {
        LONG: [],
      },
      {
        LONG: [],
      },
      {
        LONG: [],
      },
      {
        LONG: [],
      },
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should not log errors in console', () => {
    const wrapper = shallow(<SpotifyTopMusic />);
    const spy = jest.spyOn(global.console, 'error');
    expect(wrapper).not.toBeNull();
    expect(spy).not.toHaveBeenCalled();
  });

  it('renders correctly', () => {
    const wrapper = shallow(<SpotifyTopMusic />);
    expect(wrapper).toMatchSnapshot();
  });
});
