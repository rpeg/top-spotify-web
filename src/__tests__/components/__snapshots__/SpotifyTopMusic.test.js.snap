// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpotifyTopMusic renders correctly 1`] = `
<div
  style={
    Object {
      "padding": "2em 0 0 0",
    }
  }
>
  <div>
    <p
      className="text-xl"
      style={
        Object {
          "marginBottom": "0uu",
        }
      }
    >
      All-time
    </p>
  </div>
  <Container
    fluid={false}
  >
    <SpinnerBlock />
    <SpinnerBlock />
    <SpinnerBlock />
    <SpinnerBlock />
  </Container>
</div>
`;
